File: KoKoslan/src/java/eval/KoKoList.java
/** 
   Andrey Arguedas Espinoza
   Daniela Armas Sanchez
   Michael Chen Wang
   Kimberly Olivas Delgado
*/

package kokoslan.java.eval;

import kokoslan.java.ast.*;
import java.util.*;
import java.io.*;
import java.util.stream.IntStream;

/*  Lists of KoKoAst objects
    KoKoAst : Generates code with PrintStream,
    and can evaluate context and pass it to kokovalues
 */
public class KoKoList extends ArrayList<KoKoAst> implements KoKoAst{

    private boolean kokoNativeList = false; //To separate list of expressions and this kind of list -> []

    public KoKoList(List<KoKoAst> list, boolean nat){
        super(list); //KoKoList is ArrayList
        kokoNativeList = nat;
    }

    public KoKoList(){
        super(); //KoKoList is ArrayList
    }

    public void genCode(PrintStream out){
        if(this.size() == 0) return;
        if(this.kokoNativeList)
            out.print("[ ");
        this.get(0).genCode(out); //First KoKoAST gens code
        this.stream()
            .skip(1)              //The first does not gen
            .forEach( t -> {
                            out.print(", ");
                            t.genCode(out); //KoKoAST gens code
            });
        if(this.kokoNativeList)
            out.print("] ");
    }

    public KoKoValue eval(KoKoContext ctx){
        KoKoListValue res = new KoKoListValue();
        IntStream.range(0, this.size()).forEach( i -> {
            res.add(this.get(i).eval(ctx));
        });
        return res;
    }

    public KoKoValue eval(){ //See the upper method
        return eval(new KoKoContext());
    }
} 

File: KoKoslan/src/kotlin/eval/KoKoList.kt
/** 
   Andrey Arguedas Espinoza
   Daniela Armas Sanchez
   Michael Chen Wang
   Kimberly Olivas Delgado
*/

package kokoslan.kt.eval

import java.util.*;
import java.io.*;
import kokoslan.kt.ast.*
import kokoslan.kt.exception.*


/*  Lists of KoKoAst objects
    KoKoAst : Generates code with PrintStream,
    and can evaluate context and pass it to kokovalues
 */
class KoKoList(list : List<KoKoAst>? = null, var kokoNativeList: Boolean = false) : ArrayList<KoKoAst>(list), KoKoAst{
    
    override fun genCode(Out : PrintStream){
        if(this.size == 0) return
        if(kokoNativeList == true)
            Out.print("[ ")
        this.get(0).genCode(Out)
        this.drop(1)
            .forEach{Out.print(", "); it.genCode(Out);}
        if(kokoNativeList == true)
            Out.print("] ")
    }

    override fun eval(ctx : KoKoContext) : KoKoValue? {
        val res = KoKoListValue()
        for(i in 0 .. this.size-1) res.add(this.get(i).eval(ctx))
        return res
    }

    fun eval() : KoKoValue? {
        return eval(KoKoContext())
    }
    
} 